contracts/token/ERC721/extensions/ERC721Consecutive.sol

```diff
@@ -0,0 +1,116 @@
+        return _sequentialBurn.get(tokenId) ? address(0) : address(_sequentialOwnership.lowerLookup(uint96(tokenId)));
+    }
+
+    /**
+     * @dev Mint a batch of token of length `batchSize` for `to`.
+     *
+     * WARNING: Consecutive mint is only available during construction. ERC721 requires that any minting done after
+     * construction emits a Transfer event, which is not the case of mints performed using this function.
+     *
+     * WARNING: Consecutive mint is limited to batches of 5000 tokens. Further minting is possible from a contract's
+     * point of view but would cause indexing issues for off-chain services.
+     *
+     * Emits a {ConsecutiveTransfer} event.
+     */
+    function _mintConsecutive(address to, uint96 batchSize) internal virtual {
+        // minting a batch of size 0 is a no-op
+        if (batchSize == 0) return;
+
+        require(!Address.isContract(address(this)), "ERC721Consecutive: batch minting restricted to constructor");
+        require(to != address(0), "ERC721Consecutive: mint to the zero address");
+        require(batchSize < 5000, "ERC721Consecutive: batches too large for indexing");
+
+        uint96 first = _totalConsecutiveSupply();
+        uint96 last = first + batchSize - 1;
+
+        // hook before
+        _beforeConsecutiveTokenTransfer(address(0), to, first, last);
+
+        // push an ownership checkpoint & emit event
+        _sequentialOwnership.push(SafeCast.toUint96(last), uint160(to));
+        emit ConsecutiveTransfer(first, last, address(0), to);
+
+        // hook after
+        _afterConsecutiveTokenTransfer(address(0), to, first, last);
+    }
+
+    /**
+     * @dev See {ERC721-_mint}. Override version that restricts normal minting to after construction.
+     *
+     * Warning: Using {ERC721Consecutive} prevents using {_mint} during construction in favor of {_mintConsecutive}.
+     * After construction, {_mintConsecutive} is no longer available and {_mint} becomes available.
+     */
+    function _mint(address to, uint256 tokenId) internal virtual override {
+        require(Address.isContract(address(this)), "ERC721Consecutive: can't mint during construction");
+
+        super._mint(to, tokenId);
+        if (_sequentialBurn.get(tokenId)) {
+            _sequentialBurn.unset(tokenId);
+        }
+    }
+
+    /**
+     * @dev See {ERC721-_mint}. Override needed to avoid looking up in the sequential ownership structure for burnt
+     * token.
+     */
+    function _burn(uint256 tokenId) internal virtual override {
+        super._burn(tokenId);
+        if (tokenId <= _totalConsecutiveSupply()) {
+            _sequentialBurn.set(tokenId);
+        }
+    }
+
+    function _totalConsecutiveSupply() private view returns (uint96) {
+        uint256 length = _sequentialOwnership._checkpoints.length;
+        return length == 0 ? 0 : _sequentialOwnership._checkpoints[length - 1]._key + 1;
```

Comment:
Instead of accessing the struct member, can't we use `latest()`?

Reply:
Two things:
- Latest returns the latest value. Here we want the latest key
- We could make a `latestKey` function (its in one of the commits for this PR) and use it, but then we have another issue:

The +1 is part of the right branch, its `0` or `_sequentialOwnership._checkpoints[length - 1]._key + 1`.
If `latestKey` returns `length == 0 ? 0 : _sequentialOwnership._checkpoints[length - 1]._key`, we still need to know if 0 is the key of a checkpoint or if it comes from an empty array.

If the first batch is a batch of length 0, then there is a key 0 in the array for that token. Its very different from the array being empty

Reply:
A batch of 1 will create `tokenId = 0` ... and since that is the last token of the batch, that is what we put in the checkpoint ... so there will be `0` in the history.

If we want `latest()` to return `(key,value)` ... that is a bit harder for dev to work with because you need to allocate the variable before using one. Also we can't change `History.latest()` for backward compatibility ... and I'd rather not have an inconsistency between `TraceXXX.latest()` and `History.latest()`.

Reply:
Proposal:

- Add `trace.latestCheckpoint()` returning `(key, value)`
- Add `trace.length()`
- Implement this using the above functions

The reason I insist is that if we have to reach for the struct's internal variables it's because the interface is not enough. So I think it's worth adding these functions for users that may need to do similar things.

Reply:
I think we already have `trace.length()`

Reply:
about `trace.latestCheckpoint()`, should it return `(key, value)` or `(pos, key, value)` ?
What if the Trace is empty, should it revert ?